d2/src/main/java/com/linkedin/d2/balancer/strategies/relative/RelativeStateUpdater.java

```diff
@@ -0,0 +1,217 @@
+    // Step 3: Handle quarantine and recovery for all tracker clients in this cluster
+    // this will adjust the base health score if there is any change in quarantine and recovery map
+    _quarantineManager.updateQuarantineState(_partitionLoadBalancerStateMap.get(partitionId));
+    // Step 4: TODO Calculate the new ring for each partition
+  }
+
+  /**
+   * Update the health score of all tracker clients for the service
+   * @param trackerClients All the tracker clients in this cluster
+   */
+  private void updateHealthScoreAndState(Set<TrackerClient> trackerClients, int partitionId, Optional<Long> maybeClusterGenerationId)
+  {
+    PartitionLoadBalancerState partitionLoadBalancerState = _partitionLoadBalancerStateMap.get(partitionId);
+    ConcurrentMap<TrackerClient, TrackerClientState> trackerClientStateMap = partitionLoadBalancerState.getTrackerClientStateMap();
+
+    // Remove the trackerClients from original map if there is any change in uri list
+    if (maybeClusterGenerationId.isPresent())
+    {
+      for (TrackerClient trackerClient : trackerClientStateMap.keySet())
+      {
+        if (!trackerClients.contains(trackerClient))
+        {
+          trackerClientStateMap.remove(trackerClient);
+          partitionLoadBalancerState.getQuarantineMap().remove(trackerClient);
+          partitionLoadBalancerState.getQuarantineHistory().remove(trackerClient);
+          partitionLoadBalancerState.getRecoveryMap().remove(trackerClient);
+        }
+      }
+    }
+
+    // Calculate the base health score before we override them when handling the quarantine and recovery
+    calculateBaseHealthScore(trackerClients, partitionId);
+
+    // Update cluster generation id if it's changed
+    if (maybeClusterGenerationId.isPresent())
+    {
+      _partitionLoadBalancerStateMap.get(partitionId).setClusterGenerationId(maybeClusterGenerationId.get());
+    }
+  }
+
+  private void calculateBaseHealthScore(Set<TrackerClient> trackerClients, int partitionId)
+  {
+    // Snap stats for each tracker client, we want to get snap the stats because they can change any time during we calculate the new health score
+    long sumAvgLatency = 0;
+    PartitionLoadBalancerState partitionLoadBalancerState = _partitionLoadBalancerStateMap.get(partitionId);
+    ConcurrentMap<TrackerClient, TrackerClientState> trackerClientStateMap = partitionLoadBalancerState.getTrackerClientStateMap();
+    Map<TrackerClient, CallTracker.CallStats> latestCallStatsMap = new HashMap<>();
+
+    for (TrackerClient trackerClient : trackerClients)
+    {
+      CallTracker.CallStats latestCallStats = trackerClient.getLatestCallStats();
+      latestCallStatsMap.put(trackerClient, latestCallStats);
+
+      sumAvgLatency += Math.round(latestCallStats.getCallTimeStats().getAverage());
+    }
+    long clusterAvgLatency = sumAvgLatency / trackerClients.size();
```

Comment:
Calculating cluster avg latency is not this simple. See rfc for formula.

Reply:
Fixed